src/rdma.c

```diff
@@ -1689,25 +1697,21 @@ static int rdmaProcessPendingData(void) {
     listNode *ln;
     rdma_connection *rdma_conn;
     connection *conn;
-    listNode *node;
     int processed;
 
     processed = listLength(pending_list);
     listRewind(pending_list, &li);
     while ((ln = listNext(&li))) {
         rdma_conn = listNodeValue(ln);
         conn = &rdma_conn->c;
-        node = rdma_conn->pending_list_node;
 
         /* a connection can be disconnected by remote peer, CM event mark state as CONN_STATE_CLOSED, kick connection
          * read/write handler to close connection */
         if (conn->state == CONN_STATE_ERROR || conn->state == CONN_STATE_CLOSED) {
-            listDelNode(pending_list, node);
-            /* do NOT call callHandler(conn, conn->read_handler) here, conn is freed in handler! */
-            if (conn->read_handler) {
-                conn->read_handler(conn);
-            } else if (conn->write_handler) {
-                conn->write_handler(conn);
+            listDelNode(pending_list, rdma_conn->pending_list_node);
+            rdma_conn->pending_list_node = NULL;
+            if (callHandler(conn, conn->read_handler)) {
+                callHandler(conn, conn->write_handler);
             }
```

Comment:
I am not familiar with the `process_pending_data` contract but I noticed a behavior change here. We are now calling both the `read_handler` and the `write_handler` while previously we invoked just one and if both are set we give the preference to `read_handler`. Is this change intentional?

Reply:
Yes, it's intentional. Please take a look at `callHandler`, if connection is closed successfully, this function returns 0, it would not invoke the second handler.
```
static inline int callHandler(connection *conn, ConnectionCallbackFunc handler) {
    connIncrRefs(conn);
    if (handler) handler(conn);
    connDecrRefs(conn);
    if (conn->flags & CONN_FLAG_CLOSE_SCHEDULED) {
        if (!connHasRefs(conn)) connClose(conn);
        return 0;
    } 
    return 1;
}
```
Oh, I think I need to add the necessary comment in code:
```
            /* in case read_handler is NULL, we have a chance to invoke write_handler.
               if the connection is closed successfully in read_handler, the returned 0
               will not invoke write_handler any more. */
```